### ISSUE TITLE
remove anonymous user

### ISSUE THREAD

**Issue by davidism** [maxcountryman/flask-login#698]
This is a constant source of confusion for users. They don't understand why `is_authenticated` is always true for `UserMixin` even if it's not the logged in user. They get tripped up when they try to access attributes on `current_user` that aren't available on `AnonymousUserMixin`.

Whenever I'm writing my own quick login implementation, I set `g.user` to `None` if there's no logged in user, and the `current_user` proxy would be unbound (`if current_user` would look `False`).

You can return any object you want from `user_loader`. If an application needs an anonymous user system, they can return an anonymous user object from `user_loader`. They'll most likely need that anyway as even anonymous users might take actions that need to be associated together, and a single anonymous user wouldn't work for that.

I think removing this might also make it more obvious that you can return more than one type from `user_loader`, like `AdminUser` and `RegularUser` for example.

I realize this is a larger departure from current behavior than other refactors we've been doing. But I think it's worth doing for a simpler API.

**Comment by maxcountryman** [comment]
Agree; this wouldn't technically remove any functionality, but it would reduce cognitive overhead. I'm in favor of that. It's probably worth documenting how anonymous users can be implemented via `user_loader` if we go this route.

**Comment by davidism** [comment]
I don't care about breaking changes in this case. I'm removing `AnonymousUserMixin`, possibly with a deprecation warning first if I can figure out how to go about that.

**Comment by davidism** [comment]
I haven't experimented yet, but my current idea to deprecate this will be to have a `LOGIN_NO_ANONYMOUS` config at first. If enabled, it will cause `current_user` to be `None` when not logged in. The `is_anonymous` and `is_authenticated` properties should warn something like "These properties are not intuitive and will go away, enable LOGIN_NO_ANONYMOUS and check 'not current_user' instead." Subclassing `AnonymousUserMixin` can also show a warning.

**Comment by jwag956** [comment]
I happened to be working getting Flask-Security-Too to be compatible with this change. As a test I set:
`login_manager.anonymous_user = lambda: None`

and started fixing issues :-)
This was pretty simple and I easily got my many many unit tests to pass. My current idea is to default the next release to do this with a backwards compat configuration.

I never realized that is_authenticated actually looks at is_active - that might cause very subtle bugs.. need to work on that next.

**Comment by davidism** [comment]
It didn't prior to 0.6. There was an issue, a PR, and then someone pointed out that it's sort of confusing now. I'd really like to simplify things like that, it seems like `is_active` is something that should be checked or not in `user_loader`, not automatically checked or even provided by `UserMixin`.

**Comment by grosscol** [comment]
I concur about removing `AnonymousUser` as I don't have a case for a singleton anonymous user.  In those instances, I would disable login for the application altogether. E.g.

```py
# auth_routes.py
def authentication_required():
    """
    This serves as the before request function when all routes in blueprint should require auth.
    """
    if request.method == 'OPTIONS':
        return None
    elif current_app.config.get('LOGIN_DISABLED'):
        return None
    elif current_user.is_authenticated:
        return None
    else:
        resp = {'message': 'Authentication required.'}
        return(make_response(jsonify(resp), 401))
```

```py
# in create_app() of  __init__.py

    # Protect data endpoint blueprints without modifying them with before_request.
    app.before_request_funcs = {
        'variant_routes': [auth_routes.authentication_required],
        'region_routes': [auth_routes.authentication_required],
    }
```

I assume there are use cases for tracking each user as a transient anonymous user that is only identified by their ip, session hash, or something of the sort. The `AnonymousUserMixin` as it exists doesn't facilitate those use cases.
